### ISSUE TITLE
Refactor EE site db schema

### ISSUE THREAD

**Issue by kirtangajjar** [EasyEngine/easyengine#1153]
Current **site** table in sqlite db looks like - 

| Column           | Type     | Constraint  | Purpose                                                               | sample values                      |
|------------------|----------|-------------|-----------------------------------------------------------------------|------------------------------------|
| id               | INTEGER  | PRIMARY KEY | auto increment mysql id                                               | 1,2,3... natural numbers. int only |
| sitename         | VARCHAR  | UNIQUE      | Domain Name of site                                                   | example.test                       |
| site_type        | VARCHAR  | -           | (package-specific) Type of site from that package.                    | wpsubdir                           |
| site_command     | VARCHAR  | -           | Type of site                                                          | wp                                 |
| site_title       | VARCHAR  | -           | (wp-specific) Site title                                              | example's site                     |
| proxy_type       | VARCHAR  | -           | Type of proxy used.                                                   | ee-nginx-proxy                     |
| cache_type       | VARCHAR  | -           | Type of cache                                                         | redis                              |
| site_path        | VARCHAR  | -           | Path on filesystem where site is stored                               | /home/user/ee-sites/example.com    |
| created_on       | DATETIME | -           | To know when site was created                                         | 2018-08-06 12:43:30                |
| is_enabled       | BOOLEAN  | -           | Keeps track of weather site is enabled                                | 1                                  |
| is_ssl           | BOOLEAN  | -           | Keeps track of ssl on site                                            | 0                                  |
| storage_fs       | VARCHAR  | -           | Carried forward from v3 in early stages.                              |                                    |
| storage_db       | VARCHAR  | -           | Carried forward from v3 in early stages.                              |                                    |
| db_name          | VARCHAR  | -           | (wp-php-specific) database name                                       | wordpress                          |
| db_user          | VARCHAR  | -           | (wp-php-specific) database user                                       | mysql                              |
| db_password      | VARCHAR  | -           | (wp-php-specific) database password                                   | 23fsd435dfg                        |
| db_root_password | VARCHAR  | -           | (wp-php-specific) database root password                              | poio54347mk                        |
| db_host          | VARCHAR  | -           | (wp-php-specific) database host                                       | mysql                              |
| db_port          | VARCHAR  | -           | database port                                                         | 3345                               |
| wp_user          | VARCHAR  | -           | (wp-specific) Store WordPress Username                                | admin                              |
| wp_pass          | VARCHAR  | -           | (wp-specific) Store WordPress User Password                           | Zg7YL72oN0R8                       |
| email            | VARCHAR  | -           | (wp-specific) Wordpress user email                                    |                                    |
| php_version      | VARCHAR  | -           | (php-specific) Version of php used by site (will be needed in future) | 7.2                                |

It was designed when we were focused on getting wp site running and so there are many wordpress specific columns in it(i.e. wp_user, wp_pass, email, db_*). Other site types like html etc will not need them. 

Another drawback is that when third party developers start to develop their package, there's no way to add additional information about the site(i.e. a nodejs site type package might need to store node version for each site).

Hence the proposal is to keep all columns which will be common in all sites in one table and separate out the rest.

So there will be two tables - **site** and **site_meta**

**site** table will look like:

| Column               | Type     | Constraint  | Values                                              |
|----------------------|----------|-------------|-----------------------------------------------------|
| site_id              | INTEGER  | PRIMARY KEY | 1,2,3...                                            |
| site_url             | VARCHAR  | UNIQUE      | foo.com, example.com, bar.example.com, foo.com/blog |
| site_type            | VARCHAR  | -           | html, wp, php                                       |
| site_fs_path         | VARCHAR  | -           | /var/ww/foo.com/                                    |
| site_enabled         | BOOLEAN  | -           | yes/no                                              |
| site_ssl             | VARCHAR  | -           | no,lets-encrypt,self-signed,csr,wildcard            |
| nginx_browser_cache  | VARCHAR  | -           |                                                     |
| nginx_fullpage_cache | VARCHAR  | -           |                                                     |
| php_version          | VARCHAR  | -           | 7.2                                                 |
| php_opcache          | VARCHAR  | -           |                                                     |
| db_name              | VARCHAR  | UNIQUE      | db_site_name                                        |
| db_user              | VARCHAR  | UNIQUE      | user_site_name                                      |
| db_password          | VARCHAR  | -           | 23fsd435dfg                                         |
| db_root_password     | VARCHAR  | -           | poio54347mk                                         |
| db_host              | VARCHAR  | -           | localhost, mysql, s3                                |
| db_port              | VARCHAR  | -           | 3345                                                |
| db_cache             | VARCHAR  | -           |                                                     |
| app_admin_url        | VARCHAR  | -           |                                                     |
| app_admin_email      | VARCHAR  | -           |                                                     |
| app_admin_username   | VARCHAR  | -           |                                                     |
| app_admin_password   | VARCHAR  | -           |                                                     |
| app_object_cache     | VARCHAR  | -           |                                                     |
| app_mail             | VARCHAR  | -           | mailhog, postfix, ses, mandrill, gmail, NULL        |
| admin_tools          | VARCHAR  | -           | enabled, disabled                                   |
| created_on           | DATETIME | -           |                                                     |
| modified_on          | DATETIME | -           |                                                     |
| site_auth_scope      | VARCHAR  | -           | complete, admin-tools                               |
| site_auth_username   | VARCHAR  | -           |                                                     |
| site_auth_password   | VARCHAR  | -           |                                                     |

**site_meta** table will look like:

| Column        | Type     | Constraint   |
|---------------|----------|--------------|
| id            | INTEGER  | PRIMARY KEY  |
| site_id       | INTEGER  | FK: sites:id |
| meta_key      | VARCHAR  |              |
| meta_value    | VARCHAR  |              |

Now all the site type specific information can go in `site_meta` table.

P.S. Updated new tables as per discussion from [spreadsheet](https://docs.google.com/spreadsheets/d/1HTl-UzUhLVl8S69870J0ooM1s44kdZ6Bzt9d0VHvMgs/)

**Comment by rahul286** [comment]
@kirtangajjar can you please move this to a Google spreadsheet so we can discuss each item one-by-one?

**Comment by kirtangajjar** [comment]
Here's the URL - https://docs.google.com/spreadsheets/d/1HTl-UzUhLVl8S69870J0ooM1s44kdZ6Bzt9d0VHvMgs/

**Comment by kirtangajjar** [comment]
Update on the issue: When the issue started, original plan was to de-normalize table and move many columns in `site_meta` table.However we realized that *most* of the site will need this sort of information(i.e. admin_email, admin_password etc...) and so we decided to keep them in the sites table itself and added other fields ee would need in it. Also this would reduce implementation complexity so this is the path we chose to go forward with.

As we supports DB migration, we can make changes to this schema later on without any issues.
